video/out/vo_vaapi.c

```diff
@@ -684,6 +684,9 @@ static void draw_osd(struct vo *vo)
         int rw = mp_rect_w(*rc);
         int rh = mp_rect_h(*rc);
 
+        if (n == num_mod_rc - 1)
+            rw = w - rc->x0;
```

Comment:
This appears to fix the buffer overflow, so thanks that's definitely an improvement. However, the dimensions of the OSD are wrong now. If you resize the video a bit with the mouse you should see it. The osc gets smaller than the window. In `vo_dmabuf_wayland`, I had to make a similar adjustment but unfortunately it doesn't seem to work here.

Reply:
I don't see any issue when changing window size. Are you on Intel? Intel supports VA_SUBPICTURE_DESTINATION_IS_SCREEN_COORD (draw subpictures in screen resolution rather that video resolution), while Mesa does not. Perhaps that makes a difference. I tried copying your code from vo_dmabuf_wayland, in fact that's what I started with, but then the bottom-right part of the OSD/OSC is not drawn, as if the last slice is simply missing.

Reply:
This vo doesn't even work on my intel (Kaby Lake) for some reason. `vaPutSurface()` fails constantly. But my main machine is AMD not intel, so maybe that matters. Does the default osc not get much smaller than the window for you if you resize it a bit?

Reply:
>LIBVA_DRI3_DISABLE=1

Oh okay so if I use this the VO works except now I don't get any osd at all for some reason. The same happens with master though.

>--rh

I tried this out of curiosity and it gives me exactly the same behavior (shrinking osc that doesn't fill the whole window).

Reply:
I just learned that the new iHD driver doesn't support subpictures at all: https://github.com/intel/media-driver/issues/1268
```
vo/vaapi] 0 subpicture formats available:
[vo/vaapi] OSD format not supported. Disabling OSD.
```
The old i965 driver does:
```
[vo/vaapi] 6 subpicture formats available:
[vo/vaapi]   AI44, flags 0x6
[vo/vaapi]   IA44, flags 0x6
[vo/vaapi]   IA88, flags 0x6
[vo/vaapi]   AI88, flags 0x6
[vo/vaapi]   BGRA, flags 0x6
[vo/vaapi]   RGBA, flags 0x6
```
As does mesa:
```
[vo/vaapi] 1 subpicture formats available:
[vo/vaapi]   BGRA, flags 0x0
```

Oh well. Time to retire (aka remove) vo_vaapi from mpv?

Reply:
Well I mean probably nobody should ever use this VO so I won't exactly miss it, but there's still plenty of people on drivers that could use it so maybe it's worth keeping around at least for another release.

As for your fix, I double checked the coordinates and I'm convinced it's right. The math adds up correctly and everything. Whatever I'm experiencing must be some kind of driver bug (maybe the `VA_SUBPICTURE_DESTINATION_IS_SCREEN_COORD` thing like you mentioned). Anyways, considering this takes me from "instant crash on resize" to "works but with a funny osd", that's still a considerable upgrade. There's already a warning that the VO has problems with the osd anyway, so I don't think I'll force you to try and debug legacy garbage in drivers.